### ISSUE TITLE
Switch to simpler looking loading spinner

### ISSUE THREAD

**Issue by jancborchardt** [owncloud/core#19091]
We should switch to a simpler spinner for the next version. The current one is very complex, has many parts, and the GIF is not optimized for HiDPI.

Similar to this maybe (of course better quality and scalable etc.):
![](http://www.designbrooklyn.com/images/loading_icon1.gif)
Or the one of the new HTML5 player on https://twitch.tv

Ref https://github.com/owncloud/gallery/issues/336 for previous discussion on the spinner. cc @owncloud/designers @oparoz

**Comment by oparoz** [comment]
GIFs don't work well on all backgrounds and should never be used for a spinner where you don't control the environment It would be best to go with CSS.

I've never been a fan of the Android spinner, but I know picking a good one is time consuming.

**Comment by oparoz** [comment]
A few more issues with GIFs: apps and users can't theme them. With CSS, you change the colour and you're in business.

**Comment by jancborchardt** [comment]
@oparoz as said in another issue: GIFs can be themed – simply change the image in the theme.

It is correct though that there are issues with the background and the look, I don’t disagree with that. However, we should approach this properly with 9.0 and from core out, not Gallery going its own way for 8.2, essentially breaking themed loading spinners.

**Comment by oparoz** [comment]
> GIFs can be themed – simply change the image in the theme

Yes, but that means creating an entirely new spinner or having excellent Photoshop skills. You can't just set the colour via CSS.

**Comment by PVince81** [comment]
There are also pure-CSS spinners (probably not working in IE8 but who cares)

**Comment by oparoz** [comment]
> probably not working in IE8 but who cares

You know who ;), Chinese users stuck on XP :)

**Comment by oparoz** [comment]
And there is a big problem which needs to be solved in core for 9.0 regarding the spinner. They need to be applied to their own divs so that themes can't use high quality spinner. You have to get rid of all the spinners which are simply applied to large containers.

**Comment by oparoz** [comment]
Use a DIV spinner, go to settings, make the window narrow, try to use the top-right menu. It doesn't work.
Just one of the examples of things which are currently broken when using a non-GIF spinner.

**Comment by jancborchardt** [comment]
Played around a bit with a good-looking CSS spinner, here’s what we could start with:
```
.loading {
	position: relative;
	display: block;
	width: 48px;
	height: 48px;
	margin: 0 auto;
	border: 3px solid rgba(240, 240, 240, .25);
	border-left-color: rgba(180, 180, 180, .5);
	border-radius: 50%;
	-webkit-animation: loading-spin .5s infinite linear;
	animation: loading-spin .5s infinite linear;
}

@-webkit-keyframes loading-spin {
	0% {
		-webkit-transform: rotate(0);
		transform: rotate(0);
	}
	100% {
		-webkit-transform: rotate(360deg);
		transform: rotate(360deg);
	}
}
@keyframes loading-spin {
	0% {
		-webkit-transform: rotate(0);
		transform: rotate(0);
	}
	100% {
		-webkit-transform: rotate(360deg);
		transform: rotate(360deg);
	}
}
```

Anyone up for playing around with that @owncloud/designers?

**Comment by oparoz** [comment]
Looking forward to testing this :)

**Comment by MorrisJobke** [comment]
I would double the time (otherwise it looks a bit too fast). Beside that it looks really nice :)

**Comment by blackcrack** [comment]
should be able to have all in the themefolder 
/theme/example/loader/loader.html
/theme/example/loader/loader.css
http://cssload.net/
to have a possible for change it easy out.
with other css designs.

best regards
Blacky

**Comment by Henni** [comment]
@jancborchardt I like your code sample, but would maybe slow it down like @MorrisJobke suggested.
This is what it looks like with an animation speed of 0.8 instead of 0.5:
http://codepen.io/Henni/pen/EPWOXK

**Comment by blackcrack** [comment]
well, i did not know Jan's issue before, but i have send also an an sample ....
https://github.com/owncloud/core/issues/21477#issuecomment-169255251
and i guess, it's maybe nice possible's.. for replace a "heavy" gif or animated picture, because with png's it's possible too (animated png's) but with css3 it's just more clean and maybe also more nice.
and the speed can be adjusting by owner in the css-file and can be at first more slow .. imho, but the 
themefile for it should be as a single file in htm (include ('$theme/loader/loader.htm') and in the cssfile would also well placed by side of the htm, as loader.css and should be by side of the theme as single possible imo, for fast find and easy change and set up ..
best regards
Blacky

**Comment by oparoz** [comment]
@blackcrack - Just add the CSS to your custom theme' CSS and it will replace the original spinner

**Comment by blackcrack** [comment]
well, some designer it's maybe happy if he mus not search in the whole css files or something and it is clean to see whitch it is which .. i can, but not any it's so well like you and me where can do it.. by thhe way, i am do not only css i am also in php  ;) but at moment i am in work with my laptop and this writing it is wat i can only do.. you see ;)
best regards
Blacky

**Comment by dominicrico** [comment]
Also would be good to put this into a pseudo element with an overlaying background so you can keep the original content of the element if you refresh or reload something via ajax.

**Comment by rullzer** [comment]
Would also be good for #22350

**Comment by prastut** [comment]
Catching up with the material trend http://codepen.io/prastutkumar/pen/grbKZZ
A little skeptical about it though. Opinions?

**Comment by jancborchardt** [comment]
@prastut the Material Design spinner takes way too much attention because it’s so complex. A spinner is not supposed to do that. Also, it’s heavily branded and associated with Google/Android, so we will not use it.

If you’re up for improving the spinner, check out the code I posted at https://github.com/owncloud/core/issues/19091#issuecomment-169295371 and feel free to implement it. :)

**Comment by prastut** [comment]
@jancborchardt I suspected that only. :D. Can I implement the above code and do a PR?

**Comment by blackcrack** [comment]
this is why i told also we should habe an extra possible for php-code like("spinner.php") and an spinner.php in the themefolder for have a faster possible to implement an other spinner in the theme-folder for make it able for have any theme his own spinner " include ($themefolder . 'spinner.php') "
and in the head the spinner.css file too with maybe "<?php print "$themefolder/spinner.css"; ?>"
and therwith if the $themefolder variable exist it's able to use ever the theme folder of the actual theme :)

best regards
Blacky

**Comment by skjnldsv** [comment]
I love the material design too, but I think @jancborchardt is right, it's too hypnotizing! :D
![tumblr_mfpszbk6rq1qmxhobo1_500](https://cloud.githubusercontent.com/assets/14975046/14616238/cae87496-05a9-11e6-94bc-1bd4bb9f46fa.gif)


The css one is better (with @Henni speed adjust) :)

**Comment by jancborchardt** [comment]
@skjnldsv you want to take a stab at implementing it? The only worry I have is that we need to consistently replace it across ownCloud, where we currently use a gif.

Or in the first pass we create a new, high-res, half-transparent gif from that code which can work on any background. Then second step is to make it with CSS.

**Comment by skjnldsv** [comment]
What do you mean replace across owncloud? Isn't all app using the core css values?

**Comment by oparoz** [comment]
@skjnldsv - It's not that simple. I found that you can't just replace a background gif with some CSS, you usually need to wrap the loader and that can lead to all sorts of issues, but maybe it's only with more complex loaders

**Comment by skjnldsv** [comment]
Depends, but with some :after and :before you should be able to do something simple.
I will try. :)

**Comment by skjnldsv** [comment]
For example: http://codepen.io/skjnldsv/pen/grKpOo
With `::after` I was able to separate the two section giving a beautiful loading spinner!

This shouldn't be a problem if we replaced it in the css I think. I'm gonna try.

**Comment by skjnldsv** [comment]
What about animated svg?
We could set it in background and rescaling them would be good looking?

**Comment by jancborchardt** [comment]
Oooh nice! If this works across the browsers we need to support then that would be a great solution!

**Comment by skjnldsv** [comment]
Linux:
- Chome 50: ok
- Firefox 45: ok
- Stock android 4.4.4: ok

**Comment by jancborchardt** [comment]
Important is also Internet Explorer. @MorrisJobke what was the support for 9.1 again – IE 10/11?

**Comment by oparoz** [comment]
@skjnldsv - Please create a PR :). That's the easiest thing to test. 

```
.loading,
.icon-loading,
.icon-loading-dark {
	your CSS
}
```

**Comment by skjnldsv** [comment]
Well, it doesn't seem to work on ie, but it's not coming form the principe of having css inside a svg, because http://codepen.io/skjnldsv/pen/oxyjoQ is working.
Can someone help with the animation, and figure out why this isn't working?

**Comment by oparoz** [comment]
Maybe you need a different prefix?

**Comment by PVince81** [comment]
@jancborchardt IE support timeline: https://github.com/owncloud/core/issues/15567#issuecomment-166754301

Seems we only need to support IE11 and Edge

**Comment by skjnldsv** [comment]
Why is it ALWAYS microsoft the problem... This is terribly frustrating!
Well, we should get back to the old gif. I can generate gifs from the svg animation.

**Comment by oparoz** [comment]
I think the best option is CSS + wrapper. All the locations where a Gif is used should be updated to support the new loader. Gifs are fugly.

**Comment by skjnldsv** [comment]
What about svg animation with ie fallback to gif?

**Comment by skjnldsv** [comment]
This seems to work pretty well!
```
.loading,
.loading-small,
.icon-loading,
.icon-loading-dark,
.icon-loading-small {
	position: relative;
}
.loading::after,
.loading-small::after,
.icon-loading::after,
.icon-loading-dark::after,
.icon-loading-small::after {
    content: '';
    position: absolute;
    background-size: 32px 32px;
    animation: loading-spin .8s infinite linear;
    -webkit-animation: loading-spin .8s infinite linear;
    background-repeat: no-repeat;
    background-position: center center;
    height: 32px;
    width: 32px;
    left: 50%;
    top: 50%;
    margin-left: -16px;
    margin-top: -16px;
}
.loading::after,
.icon-loading::after {
	background-image: url('../img/loading.svg');
}
.icon-loading-dark::after {
	background-image: url('../img/loading-dark.svg');
}
.loading-small::after,
.icon-loading-small::after {
	background-image: url('../img/loading.svg');
	background-size: 16px 16px;
}
.icon-32 {
	background-size: 32px !important;
}
@-webkit-keyframes loading-spin {
	100% { -webkit-transform: rotate(360deg); }
}
@keyframes loading-spin {
	100% { transform: rotate(360deg); }
}
```

**Comment by oparoz** [comment]
>What about svg animation with ie fallback to gif?

That works too :)

**Comment by skjnldsv** [comment]
Can you all test this with différents browsers? http://codepen.io/skjnldsv/pen/ONENbE

This is a svg hack. It seems to work with ie11 too! :D

**Comment by oparoz** [comment]
>This seems to work pretty well!

Doesn't work for the spinner which shows up in the menu when switching app.

**Comment by oparoz** [comment]
>Can you all test this with différents browsers? http://codepen.io/skjnldsv/pen/ONENbE

You can't use `id="loading"` as some divs will already have an ID

**Comment by skjnldsv** [comment]
This is just an example, i'm using the class for the loader.

**Comment by oparoz** [comment]
OK, but we still need a full PR so that we can test in all browsers. The svg solution works well (except in IE)

**Comment by skjnldsv** [comment]
Ready for testing, I don't really like this solution, seems a lot more complicated than it should be.
Can't we drop ie support? X)

**Comment by oparoz** [comment]
It worked well in non-M$ browsers and it's good that you've also included `dark-small`

**Comment by skjnldsv** [comment]
Yep, but i'm dropping this pr, not a good way to do it.
We should include a fallback, and that should be ok.

**Comment by MorrisJobke** [comment]
If the SVG doesn't work out maybe we should just move on and provide the new and the old way in parallel and then switch the old loading spinner of later (maybe with ownCloud 10 or 11)

**Comment by skjnldsv** [comment]
here is a full post about my tests and researches:
There are many ways to animate svg:
 1. Css embeded in the svg _http://codepen.io/skjnldsv/pen/aNKqbz_
 2. SMIL animations embeded in svg 
 3. Css animation outside the svg (with `::after` and `::before`) _http://codepen.io/skjnldsv/pen/ONENbE_
 4. Background-position animation with steps _http://codepen.io/skjnldsv/pen/oxypqg_

Pros and cons:
 1. Css inside svg are ok on most of browsers (except ie) but there's a bug in firefox which ignore the embeded css if linked as background-image or src. So we can't use an animated svg in ff except if the svg is hardcoded in the html.
 2. SMIL are pretty good and works almost everywhere (edge and ie11 included) Firefox too. BUT, chromium will drop the support in a near future (month? years?).
 3. ::after and ::before are a little hackish. It's not the best way to implement animation, but works most of the time. However,  it ONLY work for non "replaced-elements". So autoclosed elements like inputs, img... won't work.
4. More heavy with ressources, and also not the best way. It's like having an old photochemical film with spaced snapshots of the animation and quickly make it pass in the background. It's also not very practical to maintain as there is multiple duplicate of the same image inside the one we're using. Altough, it works with all browsers.

I suggest we go with smil. Even if chrome will drop the support eventually, for now it works, and should gives us some times before we need to change to full-css animations. However, we can also use smil AND css at the same time to be a step ahead and have a compatibility of both before AND after the smil support dropped.

**Comment by jancborchardt** [comment]
Sounds reasonable, I'd say go for it. Stuff always evolves and needs to be changed eventually, let's do what works best now.

**Comment by oparoz** [comment]
>SMIL are pretty good and works almost everywhere (edge and ie11 included) 

Not according to http://caniuse.com/#feat=svg-smil

**Comment by skjnldsv** [comment]
Ouups, my bad!
That being said, we still have the gif fallback for ie, so we should be good!

**Comment by oparoz** [comment]
Yes and http://loading.io/ can help us come up with both the SVG and the GIF

**Comment by skjnldsv** [comment]
Yep, but the gifs are terrible! :(

**Comment by oparoz** [comment]
I don't think they look that bad...

![ring-alt](https://cloud.githubusercontent.com/assets/323220/14671477/404b32c0-06f2-11e6-9e3c-8cc4acb2ac37.gif)

**Comment by skjnldsv** [comment]
^^
![capture d ecran_2016-04-20_12-21-58](https://cloud.githubusercontent.com/assets/14975046/14671525/7e4a3652-06f2-11e6-8bef-da13cf696754.png)

**Comment by oparoz** [comment]
Of course, but it's the same with today's gifs, they just look horrible on a dark background.

**Comment by oparoz** [comment]
The dark version

![ring-alt-dark](https://cloud.githubusercontent.com/assets/323220/14671561/b9ff5718-06f2-11e6-8e9c-be6d7bcc6dbe.gif)

**Comment by skjnldsv** [comment]
I know! Just found this funny! :)
Didn't mean to mock you! ;)

**Comment by oparoz** [comment]
Don't worry, I didn't take it personal ;). What I'm saying is that GIFs are always going to look bad, especially because you don't control the environment, so a shame for IE users, but not a big deal.

**Comment by MorrisJobke** [comment]
Why don't we go the full CSS way? We still could support the old way, but I like it more to have the nice and correct way of handling it (and require code changes) instead of stay with the old and clumsy way to have a nice migration but build on technology that doesn't seem to be the right way?

I would go the CSS way and still support the old background image based spinners. And after a grace period (and deprecation announcement) drop them. This would make the transition smooth ;)

**Comment by skjnldsv** [comment]
@MorrisJobke so smil + css? Css only exclude firefox, and it's still used a lot! :/
I could try to work on the fallback.

**Comment by oparoz** [comment]
>I would go the CSS way and still support the old background image based spinners

In my experience, this doesn't work, you see both at the same time.

**Comment by MorrisJobke** [comment]
> @MorrisJobke so smil + css? Css only exclude firefox, and it's still used a lot! :/
> I could try to work on the fallback.

Really? Firefox could not use any CSS transitions?



> I would go the CSS way and still support the old background image based spinners
> In my experience, this doesn't work, you see both at the same time.

Then we simply need to add them as a different class name ;) it is even a complete different concept. Fine with me.

**Comment by skjnldsv** [comment]
We're loosing focus! :)
SMIL is working on all browsers except ie.
I suggest we go for it and wait for full css support. If you look at the support chart http://caniuse.com/#feat=svg-smil, we have at least 4 future chrome and firefox version that still support it. This should gives us plenty of time to change it when the time will come.

Either way, i'm currently working on a hack to have a svg with css AND smil combined that override the smil animation in case of css support, i'm very close to success. If so, we will use it because it will assure us a full compatibility and a smooth transition from smil abandon to full css support.

**Comment by jancborchardt** [comment]
@skjnldsv yeah! Take charge and just go for it :) not so much discussion. If something doesn't work, we can still fix it going forward.

**Comment by skjnldsv** [comment]
Ok, then I close this one, let's review #24107.

I tried to mix both of them, but the firefox bug is too much.

**Comment by oparoz** [comment]
There is an alternative which would work in all browsers. Use a transparent PNG instead of a SVG.

**Comment by blackcrack** [comment]
and therefore would be pretty well, if exist an ".spinner { }" in the css and the html code a 
```css
class='spinner' 
```
to have later the possible for more easy change out the spinner in themes.

best regards
Blacky

**Comment by skjnldsv** [comment]
@oparoz and how will you make it spin?

**Comment by oparoz** [comment]
@skjnldsv CSS3 with :before?

**Comment by oparoz** [comment]
@skjnldsv I initially thought the whole element could rotate, but that won't work of course, since in a lot of case the sinner is added as a background-image, so using :before should work, but then we'll end up with the same issues you've previously described with input fields.

**Comment by skjnldsv** [comment]
:before is great, but look at https://github.com/owncloud/core/issues/19091#issuecomment-212352257
It exclude all elements that are auto-closed. SInce css doesn't allow :before or :after.

The best solution if we use aft/bef, is to use two elemnts with border-radius, and make one rotate. Best result and best compatibility.
A single rotating png isn't really beautiful. :/

**Comment by oparoz** [comment]
>we could also do have a js monitoring all .loading classes changes, and calculate/add a div that contains the loading data? I really don't like this idea, but it could work!

@skjnldsv That could work, but if using JS, why not simply use PNG sprites then?
http://codepen.io/sdholbs/pen/GyEDw

or even CSS only
https://jsfiddle.net/simurai/CGmCe/

**Comment by skjnldsv** [comment]
Because sprites are very heavy for the browser to render when animated, and a lot more bigger size! :/

**Comment by oparoz** [comment]
@skjnldsv - Reading [this thread](https://github.com/h5bp/lazyweb-requests/issues/103), I don't think performance would be too much of an issue, it's not like there are dozens of spinner rotating at the same time. I also don't think it's going to be worse than animated SVGs.

Also, this PNG example is 1.4k (paste in the browser). Even if we want to include all 12 steps, that's 5.6k.

```
data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAEAAAADACAMAAACDB5U0AAAAMFBMVEUAAAD///////////////////////////////////////////////////////////87TQQwAAAAEHRSTlMAUQwWSGr/BSODz+iYN7ioo0T1UwAABSpJREFUeNrtmouSozgMRbFlYfn9/387FoaIJFAYU7s7W5Vb3ZOkMzrIliD2JdNPf79UCOpJPMxV8ABADKC7UakUOgLolHRPvPbz7OEbYI1SxvaMO4Qwu28AqUroAdg4VwR9ArSqgNQ3c5xC+QQkVaWnLrm5Esw7ADi+obqmMczZvgGMMjyFnVKcAu4B1JOAlZbLNQUPAtDmbQaPM0nZO9qecwqutXKAtYTwaihz1FG2BB88ru8spYTtZLK7ElqqLJWOAd6HrNZS1nZkGEBrwq2EwP3ItKMhBM+IuLxpciF5D4yB5ZG7gWlwWD3nlyS8g5NZ3sJPT6oUQ0McZNiyZyWYzmXaTBR7cHzDg1eGLpoBawIhHgOq6LobwflCx0OQwV8g7El2HeHjApGeTmX1Sx9ZmiwqUsSTQrLMezHzTl5qcFAHIQwCzOvno4G6hqBOhzDpgUn85wRndNsXjtEdn+7J0DVCq1hixtMa0EV8cqXEWE4A5upqQC28HA8Brq5HgLGG19+jj0C5Ip6Og1p4QVpeIb7SkLUFmbWLjj4XXMs+rVeVGNHyn+3b2sJS6+N0DChx+/zE6BwDQCng0lTRlgy/SIczGBW8njsXTaUiInOSMpI2JHl+csWyWAFYX2pE5Ed4P+rliWBiBaRJAG19AlOntKvCSQDy+d4nxQnQDtC7wtjPoJreAbZrjSMz6GAPuLfKAk4gTZ+AKfWmoFsTfgE0V7IDIKeBAJpAlkZ9EsCAngNkODT99NNPz2W8N0/iQTbiY6JQRbejYqQPgPgKHdI5hKy/AdA2ctcC3vPgNyD2+inWhYqgTwBxfOm7rHMK8ROQb/gxGCrBvAMUxzO1bxrXnZ8ArG/b8E4pTkHtAcgJYL9/wDv5DALgEs7+YtFNJeOrZ3ge3Vp7WEtoLvyD6LPPuL6zlFJvJxMsJTz1DwTA+yWzdlMIhWF6+YdHQFf+ARWfGeFSa/xIb2e16vAPMC9JZISDNWSXf0CuJVFo2D9I60w88A9U8dm7R/4B5vi/9A/0U/8g7kQD/kEsEl9cp39wDhjwD1IUOXroH9j/3j/Q0zP/QDnUD/wD46KLatw/wMg7nmH/AB0r4ph/oJVbDu9G/YMaykJaN5r6pn+wjr7RAdHJTF74BwKIbtvXGURUtvnB3f4BOvEPsCrJoW/7B6ol0A494B8krKJJAHf9A45XkwDu+geGASCAu/4BcLyZBNDnH4gUA7QAbvsHrYQCuO0ftBIK4L5/IH0igBH/QACjklYeFxnz8w9++umR5P6TGQ6Wjfi4yFfR7SjnthhgAMhasouli/dF7wFih3QNB3jPg98A125Fd2jZedEOIF8qiH2XdU7BfQLK3O/HIKeQ3gGm3Y3vky4VEO0eYJfvA+h+/ypnr/YA5ATUJLrYlMVKKCCNBKEmkKVQJ/0jK9S0llL71sqOE0gvoypkOrtNpLTMY9HbyQRLCTfzn72AfHmjKnsfxT+oTbiWUPmZ5Y9vlTUEtcZfHmVIpj3Oi4I5vdVYERH1d35LUePcdGrKEZaGOPwPKbRwr65vWLqjBH3LHu2Ff1ATKHgKiDB+09aEuaS+c8GeoOlv9g+S22nEP3B7YZ9/cAqIOOIfOBH+/AMwqB74BzZhlRn2D0hxvDOj/oHCRWrQPzDYlOyYf4BN29e30l3/QLXsSbZoN/0DJdnLJvGOfwAKk943PT3wDwwHPPAPSFWN+gdiFgz6B3K4Qf9ABjziH0gJh/0DKfq4f8C1f+QfQEo//+Cnf1t/AFGZQwaktm6jAAAAAElFTkSuQmCC
```

Having said that, we could still go with CSS only and use a gif for inputs since we have those for IE anyway.

**Comment by skjnldsv** [comment]
Seems a good idea!
Here we go: http://codepen.io/skjnldsv/pen/qZgPQQ

**Comment by oparoz** [comment]
Looks good both on desktop and browser, no problem with speed.

**Comment by skjnldsv** [comment]
Seems good to me too! I updated with differents examples.
Will work on it tomorrow. Too tired this evening! :(

**Comment by oparoz** [comment]
Looks good, many thanks for the examples :)

**Comment by skjnldsv** [comment]
I'm zombying on the codepen until I fall asleep. i will integrate it on oc tomorrow.

**Comment by skjnldsv** [comment]
Done, please review!

**Comment by blackcrack** [comment]
@skjnldsv : it's maybe not bad, if be renamed ? into 
```css
/*spinner*/ 
.spinnerbig {} 
.spinnermedium {} 
.spinnersmall {} 
```
because, this name does more suitable ? .. also for the future, name it what it is .
and more easy for find.

best regards
Blacky

**Comment by skjnldsv** [comment]
Why? Don't really think this is that much of a deal :)
and I think `icon-loading`, `icon-loading-small`, `icon-loading-dark` and `icon-loading-small-dark` are pretty convenient names already ;)

**Comment by blackcrack** [comment]
it's spinn's and an icon is on the desk ;) *s* .. well, if you want search an specially icon, like an icon-loading instead a spinner, so, many fun *g* think about :) But, well you the Minion and the decisionmaker ;)
best regards
Blacky
